ratelimit.go

```diff
@@ -1,28 +1,89 @@
 package discordgo
 
 import (
+	"errors"
 	"net/http"
 	"strconv"
+	"strings"
 	"sync"
 	"sync/atomic"
 	"time"
 )
 
+// customRateLimit holds information for defining a custom rate limit
+type customRateLimit struct {
+	suffix   string
+	requests int
+	reset    time.Duration
+}
+
 // RateLimiter holds all ratelimit buckets
 type RateLimiter struct {
 	sync.Mutex
-	global          *int64
-	buckets         map[string]*Bucket
-	globalRateLimit time.Duration
+	global           *int64
+	buckets          map[string]*Bucket
+	globalRateLimit  time.Duration
+	customRateLimits []*customRateLimit
 }
 
 // NewRatelimiter returns a new RateLimiter
 func NewRatelimiter() *RateLimiter {
 
 	return &RateLimiter{
-		buckets: make(map[string]*Bucket),
-		global:  new(int64),
+		buckets:          make(map[string]*Bucket),
+		global:           new(int64),
+		customRateLimits: []*customRateLimit{},
+	}
+}
+
+// SetCustomRateLimit allows you to define a custom rate limit.
+//    suffix  :   Suffix of the bucket key. (ex) https://discordapp.com/api/channels/279809045819555840/messages//reactions//
+//    requests:   How many requests per reset
+//    reset   :   How long the reset timer is.
+func (r *RateLimiter) SetCustomRateLimit(suffix string, requests int, reset time.Duration) {
+	r.Lock()
+	defer r.Unlock()
+
+	for _, v := range r.customRateLimits {
+		if v.suffix == suffix {
+			v.requests = requests
+			v.reset = reset
+			return
+		}
+	}
+	r.customRateLimits = append(r.customRateLimits, &customRateLimit{
+		suffix:   suffix,
+		requests: requests,
+		reset:    reset,
+	})
+}
+
+// RemoveCustomRateLimit removes a custom ratelimit from the ratelimiter and all its buckets
+//    suffix: The suffix of the custom ratelimiter to remove
+func (r *RateLimiter) RemoveCustomRateLimit(suffix string) error {
+	r.Lock()
+	defer r.Unlock()
+
+	found := false
+	for i, v := range r.customRateLimits {
+		if v.suffix == suffix {
+			r.customRateLimits = append(r.customRateLimits[:i], r.customRateLimits[i+1:]...)
+			found = true
+			break
+		}
+	}
+
+	if !found {
+		return errors.New("err: custom rate limit not found")
 	}
+
+	for _, b := range r.buckets {
+		if b.customRateLimit != nil && b.customRateLimit.suffix == suffix {
+			b.customRateLimit = nil
```

Comment:
Race condition here, write to customRatelimit can occur at the same time as a read on L160